starboard/shared/libiamf/iamf_audio_decoder.cc

```diff
@@ -0,0 +1,382 @@
+        case 1:
+          sound_system = SOUND_SYSTEM_MONO;
+          break;
+        case 2:
+          // Stereo output.
+          sound_system = SOUND_SYSTEM_A;
+          SB_DLOG(INFO) << "Configuring IamfAudioDecoder for stereo output";
+          break;
+        case 6:
+          // 5.1 output.
+          sound_system = SOUND_SYSTEM_B;
+          SB_DLOG(INFO) << "Configuring IamfAudioDecoder for 5.1 output";
+          break;
+        case 8:
+          // 7.1 output.
+          sound_system = SOUND_SYSTEM_C;
+          SB_DLOG(INFO) << "Configuring IamfAudioDecoder for 7.1 output";
+          break;
+        default:
+          SB_NOTREACHED();
+      }
+    } else {
+      SB_DLOG(INFO) << "Defaulting to stereo output.";
+    }
+
+    error = IAMF_decoder_output_layout_set_sound_system(decoder_, sound_system);
+    if (error != IAMF_OK) {
+      SB_DLOG(ERROR)
+          << "IAMF_decoder_output_layout_set_sound_system() fails with error "
+          << ErrorCodeToString(error);
+      return false;
+    }
+  }
+
+  error = IAMF_decoder_set_pts(decoder_, 0, 90000);
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_set_pts() fails with error "
+                   << ErrorCodeToString(error);
+    return false;
+  }
+
+  error = IAMF_decoder_set_mix_presentation_id(decoder_,
+                                               reader_.mix_presentation_id());
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_set_mix_presentation_id() fails with error "
+                   << ErrorCodeToString(error);
+    return false;
+  }
+
+  error = IAMF_decoder_peak_limiter_enable(decoder_, 0);
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_peak_limiter_enable() fails with error "
+                   << ErrorCodeToString(error);
+    return false;
+  }
+
+  error = IAMF_decoder_set_normalization_loudness(decoder_, .0f);
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR)
+        << "IAMF_decoder_set_normalization_loudness() fails with error "
+        << ErrorCodeToString(error);
+    return false;
+  }
+
+  error = IAMF_decoder_configure(decoder_, reader_.config_obus().data(),
+                                 reader_.config_size(), nullptr);
+  if (error != IAMF_OK) {
+    SB_DLOG(ERROR) << "IAMF_decoder_configure() fails with error "
+                   << ErrorCodeToString(error);
+    return false;
+  }
+
+  decoder_is_configured_ = true;
+
+  return true;
+}
+
+void IamfAudioDecoder::TeardownCodec() {
+  if (is_valid()) {
+    IAMF_decoder_close(decoder_);
+    decoder_ = NULL;
+  }
+}
+
+scoped_refptr<IamfAudioDecoder::DecodedAudio> IamfAudioDecoder::Read(
+    int* samples_per_second) {
+  SB_DCHECK(BelongsToCurrentThread());
+  SB_DCHECK(output_cb_);
+  SB_DCHECK(!decoded_audios_.empty());
+
+  scoped_refptr<DecodedAudio> result;
+  if (!decoded_audios_.empty()) {
+    result = decoded_audios_.front();
+    decoded_audios_.pop();
+  }
+  *samples_per_second = reader_.sample_rate();
```

Comment:
`reader_` gets updated for every input buffer inside `DecodeInternal()`.  If sample rate can be changed in a stream, then the sample_rate() returned here could be the sample rate of a later written input buffer, instead of this particular DecodedAudio.